[lldb] Implement cache for swift typeref metadata

augusto2112 opened the pull request.

augusto2112 commented:
Leaving this as a draft for now as it depends on some upstream changes.

augusto2112 commented:
Accompanying swift patch: https://github.com/apple/swift/pull/60497

adrian-prantl commented on lldb/source/Core/CoreProperties.td:
```diff
@@ -26,6 +26,22 @@ let Definition = "modulelist" in {
+  def EnableSwiftMetadataCache: Property<"enable-swift-metadata-cache", "Boolean">,
+    Global,
+    DefaultFalse,
+    Desc<"Enable caching for swift reflection metadata in LLDB.">;
```
above `swift` is spelled `Swift`

adrian-prantl replied:
the other options spell this as `Swift`

adrian-prantl commented on lldb/source/Core/ModuleList.cpp:
```diff
@@ -188,6 +188,36 @@ bool ModuleListProperties::SetSwiftModuleLoadingMode(SwiftModuleLoadingMode mode
+      nullptr, idx, g_modulelist_properties[idx].default_uint_value != 0);
+}
+
+uint64_t ModuleListProperties::GetSwiftMetadataCacheMaxByteSize() {
```
we should write macros for these functions :-)

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntimeDynamicTypeResolution.cpp:
```diff
@@ -661,6 +664,22 @@ class LLDBTypeInfoProvider : public swift::remote::TypeInfoProvider {
   }
+
+  void
+  registerFieldDescriptors(uint64_t InfoID,
```
document hi-level what's going on here?

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.cpp:
```diff
@@ -0,0 +1,306 @@
+  policy.Interval = std::chrono::hours(1);
+  policy.MaxSizeBytes = properties.GetSwiftMetadataCacheMaxByteSize();
+  policy.Expiration =
+      std::chrono::hours(properties.GetSwiftMetadataCacheExpirationDays() * 24);
```
wow this API is convenient!

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.cpp:
```diff
@@ -0,0 +1,306 @@
+  policy.Expiration =
+      std::chrono::hours(properties.GetSwiftMetadataCacheExpirationDays() * 24);
+  static DataFileCache *g_data_file_cache =
+      new DataFileCache(ModuleList::GetGlobalModuleListProperties()
```
is this being leaked?

augusto2112 replied:
Yes, I'll change that to be an ivar. Looking at this again, we don't want this as a static, I think.

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.cpp:
```diff
@@ -0,0 +1,306 @@
+
+void SwiftMetadataCache::registerModuleWithReflectionInfoID(ModuleSP module,
+                                                            uint64_t info_id) {
+  Log *log = GetLog(LLDBLog::Types);
```
Doesn't there need to be some form of locking here?

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.cpp:
```diff
@@ -0,0 +1,306 @@
+  Log *log = GetLog(LLDBLog::Types);
+  auto it = m_info_to_module.find(info_id);
+
+  // Since this is a static singleton, we might've registered and loaded the
```
I don't understand the static singleton part of the comment? As opposed to what?

augusto2112 replied:
It means that since we may outlive `SwiftLanguageRuntime`, we may be asked to register the same module more than once (`SwiftLanguageRuntime` will call this function only once per module during its lifetime).

augusto2112 replied:
I'll remove the singleton part of the comment though as that makes it more confusing.

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.cpp:
```diff
@@ -0,0 +1,306 @@
+    }
+  }
+  
+  m_info_to_module[info_id] = {module, false};
```
m_info_to_module.insert({info_id, {module, false}};

What does the `false` do?

augusto2112 replied:
`false` means we don't have any info for this module in the cache. This is needed by `shouldParseReflectionInfoWithId` to decide whether a caller would parse info we don't have yet.

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.cpp:
```diff
@@ -0,0 +1,306 @@
+  if (cache_version != SWIFT_METADATA_CACHE_VERSION) {
+    LLDB_LOGV(log, "[SwiftMetadataCache] Cache file version mismatch for {0}.",
+              module->GetFileSpec().GetFilename());
+    index_cache->RemoveCacheFile(module_name);
```
What happens if you have two concurrent lldb sessions open from different toolchains?

augusto2112 replied:
I'll change the hash to include the lldb version.

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.cpp:
```diff
@@ -0,0 +1,306 @@
+    return;
+  }
+
+  std::string UUID = module->GetUUID().GetAsString();
```
if there isn't elsewhere, could you add a (doxygen?) comment with the layout of the cache files?

augusto2112 replied:
There is in `SwiftMetadataCache.h`, but I think you've seen that already :)

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.cpp:
```diff
@@ -0,0 +1,306 @@
+
+  llvm::SmallString<0> decompressed;
+  auto error =
+      llvm::zlib::uncompress(string_buffer, decompressed, expanded_size);
```
For consideration: Using a format that needs to be decompressed is more expensive in terms of RAM than mmapping in data that we can use directly without dirtying any memory. But it would use more disk space.

adrian-prantl replied:
Does LLVM's OnDiskHashTable work without copying the working set into RAM?

augusto2112 replied:
I think that's a good idea, but we should try to nail the format in a second patch, as this one is pretty big already.

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.cpp:
```diff
@@ -0,0 +1,306 @@
+    const auto *mangled_name = body_extractor.GetCStr(&read_offset);
+    if (!mangled_name) {
+      LLDB_LOG(log,
+               "[SwiftMetadataCache] Failed to read magled name {0} at offset "
```
typo: `mangled`

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.cpp:
```diff
@@ -0,0 +1,306 @@
+  // entirety to find out it's true size
+  uint64_t field_descriptors_size = 0;
+  for (auto _ : field_descriptors)
+    field_descriptors_size++;
```
is there no std::algorithm for this?

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.h:
```diff
@@ -0,0 +1,83 @@
+
+  /// Version number that indicates how the data is laid out.
+  /// Version 1 is encoded as follows:
+  /// A header constituting of:
```
I think it's either `A header consisting of` or `... constituting a header`.

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.h:
```diff
@@ -0,0 +1,83 @@
+  /// Version 1 is encoded as follows:
+  /// A header constituting of:
+  /// - Version number (uint16_t).
+  /// - Module signature.
```
What exactly is the module signature? And how long is it?

augusto2112 replied:
The module signature is its UUID, which varies depending on the object file format. For Mach-O that's what's specified in the `LC_UUID`, and is 16 bytes long.

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.h:
```diff
@@ -0,0 +1,83 @@
+  /// - Size of the remainder of the contents when decompressed (uint64_t).
+  /// A body that is zlib compressed containing:
+  /// - The number (N) of name/offset pairs (uint64_t).
+  /// - N pairs composed of a c-string mangled name followed by the offset
```
C strings are expensive to read because we need to scan for the NUL byte. Maybe consider
```
ULEB length
CString
```
?

augusto2112 replied:
I think we can nail down a better format in a second patch.

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.h:
```diff
@@ -0,0 +1,83 @@
+  /// - The number (N) of name/offset pairs (uint64_t).
+  /// - N pairs composed of a c-string mangled name followed by the offset
+  ///   (uint32_t) where the corresponding field descriptor can be found in
+  //    the fieldmd section.
```
Are the pairs sorted for faster reading?

augusto2112 replied:
With the current implementation I don't think that'd matter, as we parse all strings to build the map object anyway. I think we can address this in a second patch, maybe by changing the storage to use `OnDiskHashTable` instead. I didn't want to add more to this patch as it already seemed quite big.

adrian-prantl reviewed (commented):
Wow that's quite a patch :-)
Some first thoughts inline!

adrian-prantl commented:
Similarly to the clang module cache, we need to localize this cache to the test dir when running the LLDB test suite.

augusto2112 commented:
https://github.com/apple/swift/pull/60497

@swift-ci test

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.cpp:
```diff
@@ -0,0 +1,294 @@
+}
+
+void SwiftMetadataCache::loadCacheForModule(ModuleSP module, uint64_t info_id) {
+  std::lock_guard<std::recursive_mutex> guard(m_mutex);
```
Is this redundant with the lock in registerModuleWithReflectionInfoID or can this be called from other places? If not, should this be a lambda in registerModuleWithReflectionInfoID?

augusto2112 replied:
It's redundant. My thinking was to guard this just in case we start calling it from somewhere else in the future, which would be a pretty subtle logic error. I'll merge the functions.

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.cpp:
```diff
@@ -0,0 +1,294 @@
+    LLDB_LOG(log,
+             "[SwiftMetadataCache] Failed to read cached UUID for module {0}.",
+             module->GetFileSpec().GetFilename());
+    m_data_file_cache->RemoveCacheFile(module_name);
```
I would have expected RemoveCacheFile() to be called immediately before a new file is written. What's the flow if the cache is stale?
loadCacheForModule() deletes the cache, but where is a new one written?

augusto2112 replied:
The new one is written in `registerFieldDescriptors`, which is when we have new data to cache. I think it makes sense to delete the cache immediately if it's stale/corrupted. Since we're including the lldb version as part of the hash now I don't think there would be any advantages in doing that later. Doing it earlier is also nice because we clear up disk space.

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.cpp:
```diff
@@ -0,0 +1,294 @@
+  const auto *start = (const char *)header_extractor.GetData(&read_offset, 0);
+  // Create a reference to the compressed data.
+  llvm::StringRef string_buffer(start, (uint64_t)mem_buffer_up->getBufferEnd() -
+                                           (uint64_t)start);
```
membuffer_up->getBuffer() ?
https://llvm.org/doxygen/classllvm_1_1MemoryBuffer.html#a2ea075bd68f15b57e95f0771b8ba0bca

augusto2112 replied:
Do you mean `getBufferSize`? If so, we can't do that because we have to subtract the header we just read.

adrian-prantl replied:
ah, I didn't realize that start is not the start of the buffer,

adrian-prantl commented:
To make sure this didn't get lost:
- Similarly to the clang module cache, we need to localize this cache to the test dir when running the LLDB test suite.
- I think we should add a test where we turn on logging, and build foo.dylib, log that the cache is created, rebuild foo.dylib with modification, log that the cache is deleted and recreated, then scan the log file for those log messages to test the basic cache management.

augusto2112 commented:
> Similarly to the clang module cache, we need to localize this cache to the test dir when running the LLDB test suite.

I don't understand what you mean by this. Do you mean when running tests in general? We can probably leave it off in general, and turn it on only when testing this feature.

augusto2112 commented:
https://github.com/apple/swift/pull/60497

@swift-ci test

adrian-prantl commented:
> > Similarly to the clang module cache, we need to localize this cache to the test dir when running the LLDB test suite.
> 
> I don't understand what you mean by this. Do you mean when running tests in general? We can probably leave it off in general, and turn it on only when testing this feature.

Yes, something like this: https://github.com/llvm/llvm-project/blob/cd3a234fa9272f90dcfad3158cd9b51a0e643d1b/lldb/packages/Python/lldbsuite/test/lldbtest.py#L729

Turning it off and on only for the test is fine, if we can make the testsuite faster by sharing a cache for an entire test suite run, then that's great, too.

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntime.cpp:
```diff
@@ -676,9 +684,12 @@ bool SwiftLanguageRuntimeImpl::AddJitObjectFileToReflectionContext(
         return {};
       },
       likely_module_names);
+  // We don't care to register modules generated by the jit.
```
, because they will only be used by the current process.

adrian-prantl commented on lldb/source/Core/CoreProperties.td:
```diff
@@ -26,6 +26,22 @@ let Definition = "modulelist" in {
+  def SwiftMetadataCachePath: Property<"swift-metadata-cache-path", "FileSpec">,
+    Global,
+    DefaultStringValue<"">,
+    Desc<"The path to lldb's Swift reflection cache directory.">;
```
Sorry for being obnoxious, but you spelled "LLDB" and "lldb" in two different messages.

adrian-prantl commented on lldb/source/Core/ModuleList.cpp:
```diff
@@ -108,6 +108,13 @@ ModuleListProperties::ModuleListProperties() {
+  if (llvm::sys::path::cache_directory(path)) {
+    llvm::sys::path::append(path, "lldb");
+    llvm::sys::path::append(path, "SwiftMetadataCache");
+    lldbassert(SetLLDBIndexCachePath(FileSpec(path)));
```
Looks good.

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.h:
```diff
@@ -0,0 +1,84 @@
+  /// A map from reflection infos ids to a pair constituting of its
+  /// corresponding module and whether or not we've inserted the cached metadata
+  /// for that reflection info already.
+  std::unordered_map<uint64_t, std::pair<lldb::ModuleSP, bool>>
```
Why is this an unordered_map and not a std::vector<>? I would have expected the ids to be in a contiguous range?

augusto2112 replied:
There could be holes since we don't register JITed modules.

adrian-prantl replied:
Ah, so the uint64_t is not the index that shows up in `image list` but the index of refkection sections we register with the runtime. The first 100 or so elements are identical, but that's it. And during a session the list grows steadily, and there could by dylibs loaded after a JIT session.

Last question: would a uint32_t suffice?

augusto2112 replied:
Yes, I think so, I doubt we'd ever have more than 4294967295 reflection infos :)

adrian-prantl commented on lldb/test/API/lang/swift/metadata_cache/TestSwiftMetadataCache.py:
```diff
@@ -0,0 +1,121 @@
+# TestSwiftMetadataCache.py
+#
+# This source file is part of the Swift.org open source project
```
Ever since swift and llvm share the same license we no longer need to put these headers into every file.

adrian-prantl commented on lldb/test/API/lang/swift/metadata_cache/TestSwiftMetadataCache.py:
```diff
@@ -0,0 +1,121 @@
+        # Call super's setUp().
+        TestBase.setUp(self)
+
+        cache_dir = os.path.join(self.getBuildDir(), 'swift-metadata-cache')
```
fyi: this is equivalent to `self.getBuildArtifact('swift-metadata-cache')`

adrian-prantl commented on lldb/test/API/lang/swift/metadata_cache/main.swift:
```diff
@@ -0,0 +1,17 @@
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+// -----------------------------------------------------------------------------
```
same here

adrian-prantl commented on lldb/test/API/lang/swift/metadata_cache/main_modified.swift:
```diff
@@ -0,0 +1,20 @@
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+// -----------------------------------------------------------------------------
```
and here

adrian-prantl reviewed (approved):
LGTM with a few minor nits inline

augusto2112 commented:
https://github.com/apple/swift/pull/60497

@swift-ci test

augusto2112 commented:
https://github.com/apple/swift/pull/60497

@swift-ci test

augusto2112 commented:
https://github.com/apple/swift/pull/60497

@swift-ci test

augusto2112 committed:
[lldb] Implement cache for swift typeref metadata

Implement a mechanism to cache field descriptor location information.
One of the main problems of using type metadata as debug info, when in
comes to field descriptors in particular, is that there's no easy way
to find the correct type given a (string key), so a linear search over
every source of reflection information must be done. This patch
implements a cache which allows for field descriptors to be found in a
more efficient manner.

augusto2112 commented:
https://github.com/apple/swift/pull/60497

@swift-ci test

augusto2112 commented:
The new test is failing on linux because the bot doesn't have zlib installed, so the cache is disabled as well. Since I have a new patch, which doesn't require zlib, ready to be put up for review after this one is merged, I'm temporarily skipping linux, and will re-enable it on the next patch.
